src/compiler/diagnosticMessages.json

```diff
@@ -2699,6 +2699,10 @@
         "category": "Error",
         "code": 4102
     },
+    "Keyword '{0}' cannot be used in binding element.": {

```

Comment:
Technically, a keyword can be used *in* a binding element:
```ts
function f({ enum: x }: { enum: number }) {
    console.log(x);
}
f({ enum: 1 });
```
It just can't be used as the name of a local variable.
We might want to just add a `Keyword '{0}' cannot be used as a declaration name.` diagnostic and apply it to `const enum = 0;`, `class enum { }`, `function enum() {}` too, all of which currently have poor error messages.

Reply:
does only enum keyword need to check?
Sone other keyword are always use in local variable
Eg: type, symbol

Reply:
Yeah, it would be best to have the same error message for all keywords, as in `const class = 0;` or `function function() {}`.

Reply:
ok, and one point, is the parameter is local variable？ it is a very big change

Reply:
I would say that if it's only accessible in a local scope, it's technically a local variable -- that would include classes and functions too. But that would make some people think of just `var`, so you could just say "Keyword '{0}' cannot be used as a name" instead of "local variable name".

Reply:
seems a **big** break change, there are many variable named some keyword (is typescript repo)

Reply:
By "keyword" I mean things accepted by `isNonContextualKeyword`, which already aren't allowed as variable names.

Reply:
`isNonContextualKeyword ` is also a break change  
eg:

`@types/chai`:  

```ts

    interface Assertion extends LanguageChains, NumericComparison, TypeComparison {
        ...
        ok: Assertion;
        true: Assertion;
        false: Assertion;
        null: Assertion;
        undefined: Assertion;
        ...
```

`src/compiler/types`:

```ts
    export interface JSDocAugmentsTag extends JSDocTag {
        kind: SyntaxKind.JSDocAugmentsTag;
        class: ExpressionWithTypeArguments & { expression: Identifier | PropertyAccessEntityNameExpression };
    }
```

`vue`: 

``` ts
props: {
    a: {
        type: Number,
        default: 123
    }
}

```

Reply:
a `nonContextualKeyword ` is also a break change:

`@types/node/index.d.ts`
```ts
export function emitKeypressEvents(stream: NodeJS.ReadableStream, interface?: ReadLine): void;
```

`@types/mocha/index.d.ts`
```ts
setup(interface: Mocha.Interface): Mocha;
```

`@types/node/inspector.d.ts`  
```ts
this: Runtime.RemoteObject;
```

Reply:
Looks like `interface` is between `FirstFutureReservedWord` and `LastFutureReservedWord`; we could have a `isFutureReservedWord` helper (but don't currently) and then use `isNonContextualKeyword(k) && !isFutureReservedWord(k)`.

Reply:
`const enum = 0;`, `class enum { }`, `function enum() {}` 
in parser, they are not allow keyword(`enum`) in name, should we allow them? 
in technology that is required, but seems a overkill
and it will broken some other test case
could you sure they are expected behavior?😅

eg: 

```ts
function parseNameOfClassDeclarationOrExpression(): Identifier | undefined {
            return isIdentifier() && !isImplementsClause()
                ? parseIdentifierName()
                : undefined;
        }
```

after: 
```ts
function parseNameOfClassDeclarationOrExpression(): Identifier | undefined {
            return (isIdentifier() || isKeyword(token())) && !isImplementsClause()
                ? parseIdentifierName()
                : undefined;
        }

```

Reply:
It would be best to not break any test cases, just to change the error message. We shouldn't start allowing anything new either, since `const enum = 0;` fails in JS.

Reply:
it seems a big job, include some parser and grammar checker changes, i will update this pr soon, and need some review for the change😢